### PR TITLE
chore: Improve aggregate exception message

### LINKED ISSUE #13667: `MigrateUserData` raises an exception that is not very helpful
#13651 currently uses [`getUserFriendlyErrorText`](https://github.com/ankidroid/Anki-Android/blob/a06e727f6094541b763e2dc18a300edb3e17fd26/AnkiDroid/src/main/java/com/ichi2/anki/ui/windows/managespace/FileUtils.kt#L195-L203), a rather primitive but improvable way of getting at least some sort of a short but not entirely useless message from an exception, and its traceback for debug information to be copied to clipboard. `MigrateUserData` in case of a failure [throws an `AggregateException`](https://github.com/ankidroid/Anki-Android/blob/5f3a935d9c1ff35eaa469bf9b183c48de96c0487/AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/scopedstorage/migrateuserdata/MigrateUserData.kt#L532`) with a not-so-helpful message, and does not expose sub-exceptions via stack trace text.

I suggest:
  * to not use `AggregateException` (I do not understand why it is used); or
  * to change in the following way:
     * throw the sub-exception and not `AggregateException` if there's only one; or if not,
     * have a message/localized message that reads something like “5 errors, the last being: division by zero”, or maybe special-case it in `getUserFriendlyErrorText`,
     * override `printStackTrace()` to also print all sub-exception stack traces.

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description
3 requests were made in #13667

* One was tested already: `providedExceptionReturnedIfSingletonListProvided`
* One was implemented and untested. I added a test `aggregateExceptionStackTrace`
* One was not implemented, I implemented it `aggregateExceptionFriendlyErrorMessage`

## Fixes
Fixes #13667

## Approach
Define a string and a few methods

## How Has This Been Tested?
Unit test only

## Learning (optional, can help others)
Reminder about Lithuanian, as one reason we need `one` and `other`, even though `1` is not a possible value to the  parameter

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/res/values/02-strings.xml [hunk 1]
```diff
@@ -457,4 +457,12 @@
 
     <string name="media_sync_required_title">Media Sync Required</string>
     <string name="media_sync_unavailable_message">Media sync is disabled in the settings. Please sync and backup any media which hasn\'t been synced before continuing</string>
+
+    <plurals name="aggregate_exception_user_facing_message" comment="This is a summary of a collection of errors (to show a user) The first parameter is the message of the collection of errors
+     The second parameter is the number of errors the collection contains (2 or more)
+     The third parameter is the error message of the 'final' error of the collection.
+     Defined as a plural: Lithuanian defines the form of a count noun by the final digits of the number">
+        <item quantity="one">\'%1$s\' contained %2$d error, the last being: \'%3$s\'</item>
```
Comment (david-allison) [inline AnkiDroid/src/main/res/values/02-strings.xml:465]: This is a strange one. I left it untranslated as **most** languages don't need to concern themselves with `one`.

Open to alternate thoughts

Comment (david-allison) [reply AnkiDroid/src/main/res/values/02-strings.xml]: I changed my mind. Handled

File: AnkiDroid/src/main/res/values/02-strings.xml [hunk 2]
```diff
@@ -457,4 +457,12 @@
 
     <string name="media_sync_required_title">Media Sync Required</string>
     <string name="media_sync_unavailable_message">Media sync is disabled in the settings. Please sync and backup any media which hasn\'t been synced before continuing</string>
+
+    <plurals name="aggregate_exception_user_facing_message" comment="This is a summary of a collection of errors (to show a user) The first parameter is the message of the collection of errors
+     The second parameter is the number of errors the collection contains (2 or more)
+     The third parameter is the error message of the 'final' error of the collection.
+     Defined as a plural: Lithuanian defines the form of a count noun by the final digits of the number">
+        <item quantity="one">\'%1$s\' contained %2$d error: \'%3$s\'</item>
+        <item quantity="other">\'%1$s\' contained %2$d errors, the last being: \'%3$s\'</item>
```
Comment (lukstbit) [inline AnkiDroid/src/main/res/values/02-strings.xml:466]: Looking at the current code, AggregateException is being thrown using messages like "Migration failed" and "10 consecutive exceptions without progress". Wouldn't this result in strings like:

_'10 consecutive exceptions without progress` contained 2 errors, the last being: etc_

**Review by lukstbit — COMMENTED** [review]
Looking at the current code, AggregateException is being thrown using messages like "Migration failed" and "10 consecutive exceptions without progress". Wouldn't this result in strings like:

_'10 consecutive exceptions without progress' contained 2 errors, the last being: lastExceptionMessageHere_

**Comment by david-allison** [comment]
> Looking at the current code, AggregateException is being thrown using messages like "Migration failed" and "10 consecutive exceptions without progress". Wouldn't this result in strings like:
> 
> _'10 consecutive exceptions without progress' contained 2 errors, the last being: lastExceptionMessageHere_

Changed to `2 errors, the last being: '[bb]' [10 consecutive exceptions without progress]`

File: AnkiDroid/src/main/java/com/ichi2/exceptions/AggregateException.kt [hunk 3]
```diff
@@ -26,6 +28,19 @@ import java.lang.RuntimeException
...
+
+    override fun getLocalizedMessage(): String {
+        return AnkiDroidApp.instance.resources.getQuantityString(
+            R.plurals.aggregate_exception_user_facing_message,
+            exceptions.size,
+            super.message,
+            exceptions.size,
+            exceptions.last().message
+        )
+    }
```
Comment (oakkitten) [inline AnkiDroid/src/main/java/com/ichi2/exceptions/AggregateException.kt:42]: This is not going to be properly translated in all cases, if at all. I wonder if we can do something like...

```kt
interface ExceptionWithLocalizedMessage {
    fun getLocalizedMessage(context: Context): String
}
```

Also `exceptions.last().message` is not localized

**Comment by david-allison** [comment]
I don't have this level of time to allocate for a message change. 

The code was working previously. The calling code should be fixed.

**Pull request closed without merging**